# Telecom prepaid plan analytics

You work as an analyst for the telecom operator Megaline. The company offers its clients two prepaid plans, Surf and Ultimate. The commercial department wants to know which of the plans brings in more revenue in order to adjust the advertising budget.

You are going to carry out a preliminary analysis of the plans based on a relatively small client selection. You'll have the data on 500 Megaline clients: who the clients are, where they're from, which plan they use, and the number of calls they made and text messages they sent in 2018. Your job is to analyze clients' behavior and determine which prepaid plan brings in more revenue.

# Objectives

The objective of this project is to:
- Analyze clients' behavior and determine which prepaid plans brings in more revenue
- Use analytics to assists the comercial department in making adjustment in advertising budget
- Apply Statistical Data Analysis to a real-life analytical case study.

<hr>

 # Table of contents

<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ol>
        <li><a href="#open_the_data">Open the data file and study the general information</a></li>
        <li><a href="#data_preparation">Prepare the data</a></li>
        <li><a href="#make_calculations">Make calculations and add them to the table</a></li>
        <li><a href="#carry_out_eda">Carry out exploratory data analysis</a></li>
        <li><a href="#overall_conclusion">Overall conclusion</a></li>
        <li><a href="#project_completion_checklist">Project completion checklist</a></li>
    </ol>
</div>
<br>
<hr>

<div id="open_the_data">
    <h2>Open the data file and study the general information</h2> 
</div>

We require the following libraries: *numpy* and *pandas* for data preprocessing and manipulation, *matplotlib* and *seaborn* for visualization, *math* for factorial calculation, *scipy* for statistical analysis.

In [590]:
# import pandas and numpy for data preprocessing and manipulation
import numpy as np
import pandas as pd

# matplotlib for visualization
import matplotlib.pyplot as plt
%matplotlib inline

# seaborn for statistical data visualization
import seaborn as sns

# import the math and scipy packages for statistical analysis
import math as mt
from math import factorial
from scipy import stats as st

print('Project libraries has been successfully been imported!')

Project libraries has been successfully been imported!


In [591]:
# read the data
try:
    users = pd.read_csv('https://code.s3.yandex.net/datasets/megaline_users1.csv')
    calls = pd.read_csv('https://code.s3.yandex.net/datasets/megaline_calls.csv')
    messages = pd.read_csv('https://code.s3.yandex.net/datasets/megaline_messages.csv')
    internet = pd.read_csv('https://code.s3.yandex.net/datasets/megaline_internet.csv')
    plans = pd.read_csv('https://code.s3.yandex.net/datasets/megaline_plans.csv')
except:
    users = pd.read_csv('C:/Users/hotty/Desktop/Practicum by Yandex/Projects/Statistical Data Analysis/Data/megaline_users1.csv')
    calls = pd.read_csv('C:/Users/hotty/Desktop/Practicum by Yandex/Projects/Statistical Data Analysis/Data/megaline_calls.csv')
    messages = pd.read_csv('C:/Users/hotty/Desktop/Practicum by Yandex/Projects/Statistical Data Analysis/Data/megaline_messages.csv')
    internet = pd.read_csv('C:/Users/hotty/Desktop/Practicum by Yandex/Projects/Statistical Data Analysis/Data/megaline_internet.csv')
    plans = pd.read_csv('C:/Users/hotty/Desktop/Practicum by Yandex/Projects/Statistical Data Analysis/Data/megaline_plans.csv')
print('Data has been read correctly!')

Data has been read correctly!


In [592]:
# study the general information about the dataset
df_name = ['calls', 'users', 'messages', 'internet', 'plans']
df_head = [calls, users, messages, internet, plans]

for name, top_5_df in zip(df_name, df_head):
    # use the head() method to view the first 10 rows
    print('------------------------------------------------------------------------------')
    print('The head of the {} dataframe is:'.format(name)) 
    print(top_5_df.head())
    print()
    # check the structure of our dataset
    print('The structure of the {} dataframe is:'.format(name))
    top_5_df.info()
    print()
    # check shape of our dataframe
    print('The shape of our data is', top_5_df.shape, 'with {} rows and {} columns'.format(top_5_df.shape[0], top_5_df.shape[1]))
    print()

------------------------------------------------------------------------------
The head of the calls dataframe is:
         id  user_id   call_date  duration
0   1000_93     1000  2018-12-27      8.52
1  1000_145     1000  2018-12-27     13.66
2  1000_247     1000  2018-12-27     14.48
3  1000_309     1000  2018-12-28      5.76
4  1000_380     1000  2018-12-30      4.22

The structure of the calls dataframe is:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137735 entries, 0 to 137734
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   id         137735 non-null  object 
 1   user_id    137735 non-null  int64  
 2   call_date  137735 non-null  object 
 3   duration   137735 non-null  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 4.2+ MB

The shape of our data is (137735, 4) with 137735 rows and 4 columns

------------------------------------------------------------------------------
The head of the us

### Conclusion

From the general information about the dataset, we can see that the dataset contains different data with different datatype. One of the problems identified while looking through the data is that some columns have the wrong datatype. We would need to change datatypes to the right format. For instance, *call_date* in the `calls` dataframe needs to be changed to datetime, *reg_date* and *churn_date* in the `users` need to be changed to datetime, *message_date* in the `messages` needs to be changed to datetime, *session_date* in the `internet` needs to be changed to datetime. We would also need to check for missing values, carry out some data preprocessing, and data type replacement. 

<div id="data_preparation">
    <h2>Prepare the data</h2> 
</div>

#### Convert data to the necessary types

In [593]:
# check for null values
calls.isnull().sum()

id           0
user_id      0
call_date    0
duration     0
dtype: int64

First, we check for null values. We can see that there is no null values in the `calls` dataframe. We proceed to convert datatypes in the `calls` dataframe. By inspecting the `calls` data, we identified another problem with the data. Some data in the `duration` column has values greater than 0 but less than 1 minutes. For such values, we would be rounding up `duration` less than 1 minutes to 1 minutes.

In [594]:
# change the `calls_date` to datetime datatype
calls['call_date'] = pd.to_datetime(calls['call_date'], format='%Y-%m-%d %H:%M:%S', errors='raise')

In [595]:
# check and convert call duration that lasted 1 sec to 1 minutes
calls.loc[calls['duration'] <= 1, 'duration'] = 1

# rename columns
calls.columns = ['id', 'user', 'call date', 'duration']
calls.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 137735 entries, 0 to 137734
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype         
---  ------     --------------   -----         
 0   id         137735 non-null  object        
 1   user       137735 non-null  int64         
 2   call date  137735 non-null  datetime64[ns]
 3   duration   137735 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 4.2+ MB


We look at the `user` dataframe by checking for null values and change datatype in this dataframe. Another error we noticed is in the `churn_date` column where 466 values is missing. This large portion of missing values in the `churn_date` field will be transformed into either churn ='Yes' or no churn = 'No', thus it can be analyzed as a categorcal variable.

In [596]:
# check for null values
users.isna().sum()

user_id         0
first_name      0
last_name       0
age             0
city            0
reg_date        0
tariff          0
churn_date    466
dtype: int64

In [597]:
# change the `reg_date` in the `users` to datetime
users['reg_date'] = pd.to_datetime(users['reg_date'], format='%Y-%m-%d %H:%M:%S', errors='raise')

# change the `churn date` to datetime
users['churn_date'] = pd.to_datetime(users['churn_date'], format='%Y-%m-%d %H:%M:%S', errors='raise')
users.columns = ['user', 'first name', 'last name', 'age', 'city', 'subscription date', 'plan', 'churn date']

In [598]:
# create categorical variable for whether customer churn or not
users['customer churn'] = np.where(users['churn date'].isnull(), 'No', 'Yes')
users.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   user               500 non-null    int64         
 1   first name         500 non-null    object        
 2   last name          500 non-null    object        
 3   age                500 non-null    int64         
 4   city               500 non-null    object        
 5   subscription date  500 non-null    datetime64[ns]
 6   plan               500 non-null    object        
 7   churn date         34 non-null     datetime64[ns]
 8   customer churn     500 non-null    object        
dtypes: datetime64[ns](2), int64(2), object(5)
memory usage: 35.3+ KB


We look at the `messages`, `internet` and `plans`. We need to change datatype to datetime, and rename the columns.

In [599]:
# change the `message_date` in the `messages` to datetime
messages['message_date'] = pd.to_datetime(messages['message_date'], format='%Y-%m-%d %H:%M:%S', errors='raise')
messages.columns = ['id', 'user', 'message date']
messages.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76051 entries, 0 to 76050
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id            76051 non-null  object        
 1   user          76051 non-null  int64         
 2   message date  76051 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 1.7+ MB


In [600]:
# change the `session_date` in `internet` to datetime
internet['session_date'] = pd.to_datetime(internet['session_date'], format='%Y-%m-%d %H:%M:%S', errors='raise')
internet = internet.loc[internet['mb_used'] != 0] # drop `mb_used` less than 0
internet = internet.reset_index(drop=True, inplace=False)
internet.columns = ['id', 'user', 'session date', 'data used'] # rename columns
internet.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91078 entries, 0 to 91077
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   id            91078 non-null  object        
 1   user          91078 non-null  int64         
 2   session date  91078 non-null  datetime64[ns]
 3   data used     91078 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(1)
memory usage: 2.8+ MB


In [601]:
# rename `plans` columns
plans.columns = ['messages included', 'data volume per month', 'minutes included', 'monthly fee', 'price per gb', 'price per message', 'price per minute', 'plan']
plans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   messages included      2 non-null      int64  
 1   data volume per month  2 non-null      int64  
 2   minutes included       2 non-null      int64  
 3   monthly fee            2 non-null      int64  
 4   price per gb           2 non-null      int64  
 5   price per message      2 non-null      float64
 6   price per minute       2 non-null      float64
 7   plan                   2 non-null      object 
dtypes: float64(2), int64(5), object(1)
memory usage: 256.0+ bytes


#### Conclusion

We identified the problem with the datatype in earlier section. In this section, we fixed some of the problems identifies. One of the problems identified in the dataset is that the *subscription date*, *churn date* in the `users` dataframe, *call date* in the `calls` dataframe, *message date* in the `messages` dataframe, and *session date* in the `internet` dataframe are of datatype object. We converted this datatype to datetime datatype. We also rounded up individual calls for call durations less than 1 minutes as one minute. In addition, we created a new column in the `users` dataframe to show whether a customer churn or not. In the `internet` dataframe, 13,747 rows of data with 0 *data used (mb_used)* were filtered.

#### Make calculations

In [602]:
# add new features to table
calls['month'] = calls['call date'].dt.month # create `month` call was placed
internet['month'] = internet['session date'].dt.month # create session `month` 
messages['month'] = messages['message date'].dt.month # create `month` messages were sent

In [603]:
# number of calls made and minutes used per month
calls_per_month = (calls
                   .pivot_table(index='user', values='duration', aggfunc=['count', 'sum']).reset_index())
calls_per_month.columns = ['user', 'number of calls', 'total minutes used']
calls_per_month.head()

user  number of calls  total minutes used
0  1000               16              116.83
1  1001              261             1702.19
2  1002              113              797.74
3  1003              149             1073.39
4  1004              370             2688.68

In [604]:
# summary of calls made and minutes used per month
calls_per_month.describe()

user  number of calls  total minutes used
count   481.000000       481.000000          481.000000
mean   1250.636175       286.351351         1993.376029
std     144.955118       238.182107         1660.736804
min    1000.000000         2.000000            5.600000
25%    1125.000000       108.000000          773.400000
50%    1251.000000       216.000000         1496.780000
75%    1376.000000       413.000000         2814.250000
max    1499.000000      1300.000000         9214.400000

From the number of calls made per month, we can see that about 481 users made calls out of 500 users. It is reasonable to assume that 19 users did not make any call. The person with the highest call rate made 1300 calls while the person with the lowest call rate made just 2 calls. Since the mean is *greater* than the median, we expect the distribution to be skewed to the right. This means the data could contain outliers.

In [605]:
# number of text messages sent per month
text_per_month = (messages
                   .pivot_table(index='user', values='month', aggfunc='count').reset_index())
text_per_month.columns = ['user', 'messages sent']
text_per_month.head()

user  messages sent
0  1000             11
1  1001            207
2  1002             88
3  1003             50
4  1004            177

In [606]:
# summary of messages sent per month
text_per_month.describe()

user  messages sent
count   402.000000     402.000000
mean   1248.639303     189.181592
std     143.678749     192.207920
min    1000.000000       1.000000
25%    1124.250000      60.000000
50%    1247.500000     126.500000
75%    1372.750000     258.750000
max    1497.000000    1288.000000

On the average, about 189 messages was sent per user per month. The least number of message sent was 1 while the most message sent was 1288. Since the mean is *greater* than the median, we expect the distribution to be skewed to the right. This means the data could contain outliers. 

In [610]:
# volume of data per month
volume_per_month = (internet
                    .pivot_table(index='user', values='data used', aggfunc=['count', 'sum']).reset_index())
volume_per_month.columns = ['user', 'count', 'data volume'] 
volume_per_month.head()

user  count  data volume
0  1000      4      1901.47
1  1001    210     80437.94
2  1002     96     40293.33
3  1003     51     27044.14
4  1004    398    156352.81

In [612]:
# summary of volume of data per month per user
volume_per_month.describe()

user       count    data volume
count   489.000000  489.000000     489.000000
mean   1248.464213  186.253579   78610.968732
std     144.361253  139.418851   61940.374028
min    1000.000000    1.000000      92.680000
25%    1123.000000   71.000000   29401.530000
50%    1247.000000  155.000000   62354.370000
75%    1373.000000  278.000000  115737.760000
max    1499.000000  741.000000  390159.350000

From the volume of data used per month, we can see that about 489 users used data out of 500 users. It is reasonable to assume that 11 users did not use data in a month. On the average, about 78,601MB of data was used per month. The person with the lowest data volume used $\approx$93MB while the person with the most data usage used $\approx$390,159MB of data. Since the mean is *greater* than the median, we would expect the distribution to be skewed to the right. This means that the data could contain outliers. 

We proceed to join the calculated values to obtain a new dataset.

In [617]:
# plans for each user
user_plans = users[['user', 'plan']]
user_plans.head()

user      plan
0  1000  ultimate
1  1001      surf
2  1002      surf
3  1003      surf
4  1004      surf

In [ ]:
# join tables to gether using `merge()`


#### Conclusion

#### Processing duplicates

#### Conclusion

<div id="make_calculations">
    <h2>Make calculations and add them to the table</h2> 
</div>

<div id="carry_out_eda">
    <h2>Carry out exploratory data analysis</h2> 
</div>

<div id="overall_conclusion">
    <h2>Overall conclusion</h2> 
</div>